File: Frontend/angular-app/src/app/modules/accounting/repositories/account-head/commands/update-account-head-command.ts
import {Inject} from "@angular/core";
import {ServiceResult} from "../../../../../core/models/service-result";
import {IRequest, IRequestHandler} from "../../../../../core/services/mediator/interfaces";
import {HttpService} from "../../../../../core/services/http/http.service";
import {ApplicationError} from "../../../../../core/exceptions/application-error";
import {MediatorHandler} from "../../../../../core/services/mediator/decorator";
import {ValidationRule} from "../../../../../core/validation/validation-rule";
import {HttpRequest} from "../../../../../core/services/http/http-request";
import {AccountHead} from "../../../entities/account-head";
import {NotificationService} from "../../../../../shared/services/notification/notification.service";
import {AddAccountReferenceGroupToAccountHeadCommand} from "./add-account-reference-group-to-account-head-command";

export class UpdateAccountHeadCommand extends IRequest<UpdateAccountHeadCommand, AccountHead> {
  public id:number | undefined = undefined;
  public codeLevel: number | undefined = undefined;
  public code: string | undefined = undefined;
  public lastLevel: boolean | undefined = undefined;
  public parentId: number | undefined = undefined;
  public parentCode: string | undefined = undefined;
  public parentTitle: string | undefined = undefined;
  public title: string | undefined = undefined;
  public description: string | undefined = undefined;
  public balanceId: number | undefined = undefined;
  public balanceBaseId: number | undefined = undefined;
  public transferId: number | undefined = undefined;
  public groupId: number | undefined = undefined;
  public currencyBaseTypeId: number | undefined = undefined;
  public currencyFlag: boolean | undefined = undefined;
  public exchengeFlag: boolean | undefined = undefined;
  public traceFlag: boolean | undefined = undefined;
  public quantityFlag: boolean | undefined = undefined;
  public isActive: boolean | undefined = undefined;
  public groups: AddAccountReferenceGroupToAccountHeadCommand[] = [];
  public accessPermission:string | undefined = undefined;
  constructor() {
    super();
  }

  mapFrom(entity: AccountHead): UpdateAccountHeadCommand {
    this.mapBasics(entity, this);
    entity.accountHeadRelReferenceGroups.forEach(x => {
      this.groups.push(new AddAccountReferenceGroupToAccountHeadCommand().mapFrom(x))
    })
    return this;
  }

  mapTo(): AccountHead {
    throw new ApplicationError(UpdateAccountHeadCommand.name, this.mapTo.name, 'Not Implemented Yet')
  }

  get url(): string {
    return "/accounting/accountHead/update";
  }

  get validationRules(): ValidationRule[] {
    return [];
  }
}

@MediatorHandler(UpdateAccountHeadCommandHandler.name)
export class UpdateAccountHeadCommandHandler implements IRequestHandler<UpdateAccountHeadCommand, AccountHead> {
  constructor(
    @Inject(HttpService) private _httpService: HttpService,
    @Inject(NotificationService) private _notificationService: NotificationService
  ) {
  }

  async Handle(request: UpdateAccountHeadCommand): Promise<AccountHead> {
    let httpRequest: HttpRequest<UpdateAccountHeadCommand> = new HttpRequest<UpdateAccountHeadCommand>(request.url, request);
    // httpRequest.Query  += ;
    // httpRequest.Headers = httpRequest.Headers.append()

    return await this._httpService.Put<UpdateAccountHeadCommand, ServiceResult<AccountHead>>(httpRequest).toPromise().then(response => {
      this._notificationService.showSuccessMessage()
      return response.objResult
    })
  }
}

File: Backend/Microservices/Accounting/Eefa.Accounting.Application/UseCases/AccountHead/Command/Update/UpdateAccountHeadCommand.cs
using System.Collections.Generic;
using System.Linq;
using AutoMapper;
using Eefa.Accounting.Application.UseCases.AccountHead.Model;
using MediatR;
using System.Threading;
using System.Threading.Tasks;
using Library.Common;
using Library.Interfaces;
using Library.Mappings;
using Library.Models;
using Microsoft.EntityFrameworkCore;


namespace Eefa.Accounting.Application.UseCases.AccountHead.Command.Update
{
    public class UpdateAccountHeadCommand : CommandBase, IRequest<ServiceResult>, IMapFrom<UpdateAccountHeadCommand>, ICommand
    {
        public int Id { get; set; }
        public string Code { get; set; } = default!;
        public bool LastLevel { get; set; } = default!;
        public int? ParentId { get; set; }
        public string Title { get; set; } = default!;
        public int? BalanceId { get; set; }
        public int BalanceBaseId { get; set; } = default!;
        public int TransferId { get; set; } = default!;
        public int? GroupId { get; set; }
        public int CurrencyBaseTypeId { get; set; } = default!;
        public bool CurrencyFlag { get; set; } = default!;
        public bool ExchengeFlag { get; set; } = default!;
        public bool TraceFlag { get; set; } = default!;
        public bool QuantityFlag { get; set; } = default!;
        public bool? IsActive { get; set; } = default!;
        public string Description { get; set; }
        public List<AddAccountReferenceGroupToAccountHeadCommand> Groups { get; set; }


        public void Mapping(Profile profile)
        {
            profile.CreateMap<UpdateAccountHeadCommand, Data.Entities.AccountHead>();
        }
    }


    public class UpdateAccountHeadCommandHandler : IRequestHandler<UpdateAccountHeadCommand, ServiceResult>
    {
        private readonly IRepository _repository;
        private readonly IMapper _mapper;
        private readonly ICurrentUserAccessor _currentUserAccessor;
        public List<Data.Entities.AccountHead> AccountHeads = new List<Data.Entities.AccountHead>();
        public UpdateAccountHeadCommandHandler(IRepository repository, IMapper mapper, ICurrentUserAccessor currentUserAccessor)
        {
            _mapper = mapper;
            _currentUserAccessor = currentUserAccessor;
            _repository = repository;
        }

        public async Task<ServiceResult> Handle(UpdateAccountHeadCommand request, CancellationToken cancellationToken)
        {
            var entity = await _repository
                .Find<Data.Entities.AccountHead>(c =>
            c.ObjectId(request.Id))
                .Include(x => x.AccountHeadRelReferenceGroups)
            .FirstOrDefaultAsync(cancellationToken);

            if (request.Code != entity.Code)
            {
                this.AccountHeads = await _repository.GetAll<Data.Entities.AccountHead>().ToListAsync();

                await this.UpdateChildrenCode(entity.Id,entity.Code, request.Code);
            }

            _mapper.Map(request, entity);


            if (entity.AccountHeadRelReferenceGroups?.Count > 0 || request.Groups?.Count > 0)
            {
                var groupIdsToAdd = request.Groups?.Select(x => x.ReferenceGroupId).ToList().Except(entity.AccountHeadRelReferenceGroups?.Select(x => x.ReferenceGroupId).ToList()).ToList();
                var groupIdsToRemove = entity.AccountHeadRelReferenceGroups?.Select(x => x.ReferenceGroupId).ToList().Except(request.Groups?.Select(x => x.ReferenceGroupId).ToList()).ToList();

                foreach (var groupId in groupIdsToRemove)
                {
                    _repository.Delete(entity.AccountHeadRelReferenceGroups.FirstOrDefault(x => x.ReferenceGroupId == groupId));
                }
                foreach (var groupId in groupIdsToAdd)
                {
                    var newRelation = new Data.Entities.AccountHeadRelReferenceGroup
                    {
                        ReferenceGroupId = groupId,
                        ReferenceNo = request.Groups.FirstOrDefault(x => x.ReferenceGroupId == groupId).ReferenceNo
                    };
                    entity.AccountHeadRelReferenceGroups.Add(newRelation);
                    _repository.Insert(newRelation);
                }
            }
            _repository.Update(entity);


            await _repository.SaveChangesAsync(request.MenueId, cancellationToken);
            return ServiceResult.Success(_mapper.Map<AccountHeadModel>(entity));

        }

        public async Task UpdateChildrenCode(int parentId, string oldCode, string newCode)
        {
            foreach (var childAccountHead in this.AccountHeads.Where(x => x.ParentId == parentId).ToList())
            {
                var itemCode = childAccountHead.Code.Substring(oldCode.Length, (childAccountHead.Code.Length - oldCode.Length));
                childAccountHead.Code = newCode + itemCode;
                _repository.Update(childAccountHead);
                if (this.AccountHeads.Any(x => x.ParentId == childAccountHead.Id)) await UpdateChildrenCode(childAccountHead.Id, oldCode, newCode);
            }
        }
    }
}
